Improve stack usage to increase recursive initialization depth

AaronBallman opened the pull request:
We were crashing due to stack exhaustion on rather reasonable C++ template code. After some investigation, I found that we have a stack-allocated object that was huge: `InitializationSequence` was 7016 bytes. This caused an overflow with deep call stacks in initialization code.

With these change, `InitializationSequence` is now 248 bytes.

With the original code, testing RelWithDebInfo on Windows 10, all the tests in SemaCXX took about 6s 800ms. The max template depth I could reach on my machine using the code in the issue was 708. After that, I would get `-Wstack-exhausted` warnings until crashing at 976 instantiations.

With these changes on the same machine, all the tests in SemaCXX took about 6s 500ms. The max template depth I could reach was 1492. After that, I would get `-Wstack-exhausted` warnings until crashing at 2898 instantiations.

This improves the behavior of #88330 but there's still an outstanding question of why we run out of stack space and crash in some circumstances before we're able to issue a diagnostic about stack space exhaustion.

AaronBallman committed:
More efficient stack usage; NFC

AaronBallman committed:
Further improvements to stack usage

AaronBallman committed:
Add a release note

erichkeane commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -203,6 +203,10 @@ Non-comprehensive list of changes in this release
   like ``typeof_unqual`` from C23, similar to ``__typeof__`` and ``typeof``.
 
+- Improved stack usage with C++ initialization code. This allows significantly
+  more template instantiations before reaching stack exhaustion limits.
```
```suggestion
  more levels of recursive initialization before reaching stack exhaustion limits.
```

This isn't REALLY a template problem, and you could have manually created this problem, but templates make it 'easier' to hit this.  Perhaps the commit message should be improved too?

erichkeane commented on clang/lib/Sema/SemaOverload.cpp:
```diff
@@ -1057,16 +1057,15 @@ bool OverloadCandidateSet::OperatorRewriteInfo::shouldAddReversed(
     for (auto &C : i->Conversions)
-      C.~ImplicitConversionSequence();
+      C.~ImplicitConversionSequence();    
+    Ctx.Deallocate(i->Conversions.begin());
```
```suggestion
    Ctx.Deallocate(i->Conversions.data());
```

AaronBallman committed:
Fix formatting

erichkeane commented on clang/lib/Sema/SemaInit.cpp:
```diff
@@ -6113,8 +6113,10 @@ static bool TryOCLZeroOpaqueTypeInitialization(Sema &S,
-      FailedCandidateSet(Kind.getLocation(), OverloadCandidateSet::CSK_Normal) {
+    : FailedOverloadResult(OR_Success), FailedCandidateSet(nullptr) {
+  FailedCandidateSet = S.getASTContext().Allocate<OverloadCandidateSet>();
+  FailedCandidateSet = new (FailedCandidateSet) OverloadCandidateSet(
```
As a future optimization, I wonder if we could defer this allocation until we already know that we've failed?  Perhaps in `SetOverloadFailure`?  Not sure I can grok the analysis at the moment to know how possible this is though.

AaronBallman committed:
Update clang/docs/ReleaseNotes.rst

Co-authored-by: Erich Keane <ekeane@nvidia.com>

AaronBallman committed:
Update clang/lib/Sema/SemaOverload.cpp

Co-authored-by: Erich Keane <ekeane@nvidia.com>

AaronBallman committed:
More updates to the release notes

AaronBallman committed:
Merge branch 'aballman-stack-size' of https://github.com/AaronBallman/llvm-project into aballman-stack-size

AaronBallman commented:
The precommit CI failures are real; I'm still investigating what's going sideways.

AaronBallman committed:
Merge remote-tracking branch 'origin/main' into aballman-stack-size

AaronBallman committed:
Fix after merge

cor3ntin commented on clang/include/clang/Sema/Overload.h:
```diff
@@ -1163,7 +1137,7 @@ class Sema;
     ConversionSequenceList
     allocateConversionSequences(unsigned NumConversions) {
       ImplicitConversionSequence *Conversions =
-          slabAllocate<ImplicitConversionSequence>(NumConversions);
+          Ctx.Allocate<ImplicitConversionSequence>(NumConversions);
 
       // Construct the new objects.
       for (unsigned I = 0; I != NumConversions; ++I)
```
Maybe something like that would be cleaner

```cpp
 ConversionSequenceList allocateConversionSequences(unsigned NumConversions) {
      ImplicitConversionSequence *Conversions = Ctx.Allocate<ImplicitConversionSequence>(NumConversions);
      std::uninitialized_default_construct(Conversions, Conversions + NumConversions);
      return ConversionSequenceList(Conversions, NumConversions);
    }
```

mizvekov commented on clang/include/clang/Sema/Overload.h:
```diff
@@ -1070,57 +1077,24 @@ class Sema;
-    // inline to avoid allocation for small sets.
-    llvm::BumpPtrAllocator SlabAllocator;
+    ASTContext &Ctx;
+    SmallVector<OverloadCandidate, 4> Candidates;
+    llvm::SmallPtrSet<uintptr_t, 4> Functions;
```
pre-existing small nit, for consistency:
```suggestion
    SmallVector<OverloadCandidate, 4> Candidates;
    SmallPtrSet<uintptr_t, 4> Functions;
```

AaronBallman replied:
The reason it works to drop the `llvm::` prefix for `SmallVector` is because we do this: https://github.com/llvm/llvm-project/blob/main/clang/include/clang/Basic/LLVM.h

`SmallPtrSet` is not included in there which is why the prefix is necessary, and I'm not certain it's worth including it or not.

Thoughts?

mizvekov replied:
I see, it's fine either way 😃

AaronBallman committed:
Call the destructor for the allocated object

It turns out this is important for some strange reason. ;-) Many thanks
to @ErichKeane and @cor3ntin for helping me solve this!

AaronBallman committed:
Use the STL rather than do work manually; NFC

AaronBallman committed:
Merge remote-tracking branch 'origin/main' into aballman-stack-size

AaronBallman committed:
Merge remote-tracking branch 'origin/main' into aballman-stack-size

cor3ntin reviewed (approved):
This was a roller coaster!
LGTM

AaronBallman committed:
Don't arena allocate

These don't need to be long-lived and there's likely to be a lot of
overload and initialization code in a TU, so this should hopefully
reduce memory pressure from the heap allocations. It also simplifies
the patch by not needing to pass around an ASTContext.

AaronBallman committed:
Fix formatting; NFC

vitalybuka commented:
A new memory leak https://lab.llvm.org/buildbot/#/builders/5/builds/42694

nikic commented:
This change seems to cause significant compile-time regressions for C++ code (https://llvm-compile-time-tracker.com/compare.php?from=184ba038ac1d444980b3e554b0057f3f30c516ab&to=4082a7554521572a65a5a0008c4661a534df659d&stat=instructions%3Au).

Probably most damning are the times for the clang build itself (https://llvm-compile-time-tracker.com/compare_clang.php?from=184ba038ac1d444980b3e554b0057f3f30c516ab&to=4082a7554521572a65a5a0008c4661a534df659d&stat=instructions%3Au&sortBy=interestingness) where files like X86ISelDAGToDAG.cpp regress by >3% even in an optimized build.

vitalybuka commented:
have ideas about leak, but not sure what to do with perf regression and #88330

Proposing revert https://github.com/llvm/llvm-project/pull/89006

vitalybuka commented on clang/include/clang/Sema/Overload.h:
```diff
@@ -1163,12 +1136,7 @@ class Sema;
-      for (unsigned I = 0; I != NumConversions; ++I)
-        new (&Conversions[I]) ImplicitConversionSequence();
-
+          new ImplicitConversionSequence[NumConversions];
```
Seems unusual to keep ownership in ArrayRef.
Could we make ConversionSequenceList just a std::vector if it owns the memory?

cor3ntin commented:
@AaronBallman ping

AaronBallman commented:
> @AaronBallman ping

I've not been able to get back on to this, unfortunately. I'm not certain when I'll have time, either.

cor3ntin commented:
It might be worth reopening the PR and asking for help? otherwise we'll forget

*And some things that should not have been forgotten were lost.*
